torchvision/datasets/triplet.py

```diff
@@ -0,0 +1,139 @@
+
+    Args:
+        dist (torch.distributions.distribution.Distribution): A pytorch distribution
+        num_samples (int): number of non-zero samples needed
+
+    Returns:
+        list: indices of the non-zero samples
+    """
+    sample = dist.sample().nonzero().view(-1)
+    while len(sample) != num_samples:
+        sample = dist.sample().nonzero().view(-1)
+
+    return sample.tolist()
+
+
+def generate_triplets(class_samples, num_triplets):
+    """Generates a set of triplets from bins of samples
+
+    Args:
+        class_samples (list[list]): bins of samples, binned by class
+        num_triplets (int): number of triplets to be generated
+
+    Returns:
+        list[tuple]: triplets of the form (anchor, positive, negative)
+    """
+    triplets = []
+    class_dist = Multinomial(2, torch.Tensor([1.] * len(class_samples)))
+    swap = Bernoulli(torch.Tensor([0.5]))
+    for _ in range(num_triplets):
+        pos_cls, neg_cls = get_sampled_indices(class_dist, 2)
+        if swap.sample().item() == 1:
+            pos_cls, neg_cls = neg_cls, pos_cls
+
+        pos_samples, neg_samples = class_samples[pos_cls], class_samples[neg_cls]
+
+        pos_dist = Multinomial(2, torch.Tensor([1.] * len(pos_samples)))
+        neg_dist = Multinomial(1, torch.Tensor([1.] * len(neg_samples)))
+
+        anc_idx, pos_idx = get_sampled_indices(pos_dist, 2)
+        neg_idx = get_sampled_indices(neg_dist, 1)[0]
+        if swap.sample().item() == 1:
+            anc_idx, pos_idx = pos_idx, anc_idx
+
+        triplet = (pos_samples[anc_idx], pos_samples[pos_idx], neg_samples[neg_idx])
+        triplets.append(triplet)
+
+    return triplets
+
+
+class TripletDataset(data.IterableDataset, DatasetFolder):
+    """
+    A dataset with samples of the form (anchor, positive, negative), where anchor and
+    positive are samples of the same class, and negative is a sample of another class.
+    The dataset reads samples from a directory arranged in the following manner: ::
+
+        root/class_x/xxx.ext
+        root/class_x/xxy.ext
+        root/class_x/xxz.ext
+
+        root/class_y/123.ext
+        root/class_y/nsdf3.ext
+        root/class_y/asd932_.ext
+
+    Args:
+        root (string): Root directory path.
+        loader (callable): A function to load a sample given its path.
+        extensions (tuple[string]): A list of allowed extensions,
+        transform (callable, optional): A function/transform that takes in
+            a sample and returns a transformed version.
+            E.g, ``transforms.RandomCrop`` for images.
+        is_valid_file (callable, optional): A function that takes the path of an Image
+        file and checks if the file is a valid_file. Both extensions and is_valid_file
+        should not be passed.
+
+    Attributes:
+        samples (list[tuple]): List of (anchor, positive, negative) triplets
+    """
+
+    def __init__(self, root, loader, num_triplets, extensions=None, transform=None, is_valid_file=None):
```

Comment:
what about taking in the constructor a `Dataset`, which returns `(img, label)`, and enforce that it has some special structure internally, so that you can query `class_samples` and the number of classes?

This way, you would be able to use `TripletDataset` which a lot more datasets, like `Cifar10`, `MNIST`, etc

Reply:
Yes we can do this. It would also be more general. If we make this change though, I might encounter the following problem:

Let `dataset = Dataset(*args, **kwargs)`. `dataset[idx]` should return a tuple of the form `(image_path, class_index)`, instead of loading the image. This is because I create a data structure that bins these paths based on their class_idx to randomly sample from them (I use the `get_class_samples` method to do this). Can I expect this from the Dataset passed into the constructor? (The loader for the `Dataset` object would be `lambda x : x` and the loader for `TripletDataset` would be the image loader method that the user wants to use.)

If I can expect these properties from the dataset, It should be relatively simple to refactor my code to implement this. Otherwise, we will need to devise another strategy.

If this sounds good, I shall get to working on it right away.

Reply:
this is what I meant wrt having a special structure: you'd need to query the class index for a given image id.

The way I see this is to add more structure into the datasets, maybe by creating a class `ClassificationDataset` or something like that, which has methods for obtaining the info you need.

But I'm not clear on the semantics of it, so I'd rather postpone the decision to a later point.

On a similar note, what you mention about not returning the loaded image but instead the file-path is something I was thinking as well.

We could have a transform in `torchvision` that takes a file-path and returns the `Image` out of it. But this is a larger change that requires more thought

Reply:
For now I have described the semantics required from the dataset being passed in the constructor in the docstring. Will that be fine for now?

Reply:
We changed our API to support any indexable dataset as input. This has been solved.